src/ngx_http_lua_args.c

```diff
@@ -544,7 +547,122 @@ ngx_http_lua_ffi_req_get_uri_args(ngx_http_request_t *r, u_char *buf,
 
     return i;
 }
-#endif /* NGX_LUA_NO_FFI_API */
 
 
+int
+ngx_http_lua_ffi_req_get_uri_args(ngx_http_request_t *r, u_char *buf,
+    ngx_http_lua_ffi_table_elt_t *out, int count)
+{
+    int len;
+
+    len = r->args.len;
+
+    ngx_memcpy(buf, r->args.data, len);
+
+    return ngx_http_lua_ffi_req_args_helper(buf, out, count, len);
+}
+
+
+int
+ngx_http_lua_ffi_req_validate_post(ngx_http_request_t *r)
+{
+    if (r->connection->fd == (ngx_socket_t) -1) {
+        return NGX_HTTP_LUA_FFI_BAD_CONTEXT;
+    }
+
+    if (r->discard_body) {
+        return NGX_HTTP_LUA_FFI_POST_ARGS_EMPTY;
+    }
+
+    if (r->request_body == NULL) {
+        return NGX_HTTP_LUA_FFI_POST_ARGS_UNREAD;
+    }
+
+    if (r->request_body->temp_file) {
+        return NGX_HTTP_LUA_FFI_POST_ARGS_IN_FILE;
+    }
+
+    if (r->request_body->bufs == NULL) {
+        return NGX_HTTP_LUA_FFI_POST_ARGS_EMPTY;
+    }
+
+    return NGX_HTTP_LUA_FFI_POST_ARGS_OK;
+}
+
+
+int
+ngx_http_lua_ffi_req_get_post_args_len(ngx_http_request_t *r)
+{
+    size_t       len;
+    ngx_chain_t *cl;
+
+    len = 0;
+    for (cl = r->request_body->bufs; cl; cl = cl->next) {
+        len += cl->buf->last - cl->buf->pos;
+    }
+
+    return len;
+}
+
+
+int
+ngx_http_lua_ffi_req_get_post_args_count(ngx_http_request_t *r,
+    u_char *buf, int len, int max)
+{
+    int                      count;
+    u_char                  *p, *last;
+    ngx_chain_t             *cl;
+
+    if (max < 0) {
+        max = NGX_HTTP_LUA_MAX_ARGS;
+    }
+
+    p = buf;
+    for (cl = r->request_body->bufs; cl; cl = cl->next) {
+        p = ngx_copy(p, cl->buf->pos, cl->buf->last - cl->buf->pos);
+    }
+
+    last = buf + len;
+    count = 0;
+
+    for (p = buf; p != last; p++) {
+        if (*p == '&') {
+            if (count == 0) {
+                count += 2;
+
+            } else {
+                count++;
+            }
+        }
+    }
+
+    if (count) {
+        if (max > 0 && count > max) {
+            count = max;
+            ngx_log_debug1(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+                           "lua hit post args limit %d", max);
+        }
+
+        return count;
```

Comment:
Should we exit early and move this condition + return statement inside the previous loop's else, to exit the loop asap, or is it better off here?